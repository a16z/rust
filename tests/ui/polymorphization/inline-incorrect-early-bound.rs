// This test demonstrates an ICE that may occur when we try to resolve the instance
// of a impl that has different generics than the trait it's implementing. This ensures
// we first check that the args are compatible before resolving the body, just like
// we do in projection before substituting a GAT.
//
// When polymorphization is enabled, we check the optimized MIR for unused parameters.
// This will invoke the inliner, leading to this ICE.

//@ compile-flags: -Zpolymorphize=on -Zinline-mir=yes

trait Trait {
    fn foo<'a, K: 'a>(self, _: K);
}

impl Trait for () {
    #[inline]
    fn foo<K>(self, _: K) {
        //~^ ERROR lifetime parameters or bounds on method `foo` do not match the trait declaration
        todo!();
    }
}

pub fn qux<T>() {
    ().foo(());
}

fn main() {}
