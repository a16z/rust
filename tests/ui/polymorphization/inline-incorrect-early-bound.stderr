error[E0195]: lifetime parameters or bounds on method `foo` do not match the trait declaration
  --> $DIR/inline-incorrect-early-bound.rs:17:11
   |
LL |     fn foo<'a, K: 'a>(self, _: K);
   |           -----------
   |           |       |
   |           |       this bound might be missing in the impl
   |           lifetimes in impl do not match this method in trait
...
LL |     fn foo<K>(self, _: K) {
   |           ^^^ lifetimes do not match method in trait

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0195`.
